src/components/Plate/Plate.tsx

```diff
@@ -1,67 +1,101 @@
 import { classNames } from "@utils/stringUtils";
 import { SelectableGroup, createSelectable } from "react-selectable";
-import { RowAnnotation } from "@Ariadne/types";
 import { z } from "zod";
 import { wellsToRowsCols, indexToExcelCell } from "./utils";
-import { RowAnnotationGutter } from "./RowAnnotationGutter";
+import { RowAnnotationGutter, ColAnnotationGutter } from "./AnnotationGutters";
+import { AnnotationBar } from "./AnnotationBar";
+import { useState } from "react";
 
 const PlateSelectionSchema = z.object({
   wells: z.array(z.number()),
   className: z.string().optional(),
 });
 export type PlateSelection = z.infer<typeof PlateSelectionSchema>;
 
-export interface PlateProps {
+export interface WellAnnotation<T extends Record<string, unknown>> {
+  id: string;
+  wells: number[];
+  label: string;
+  className?: string;
+  metadata?: T;
+}
+export interface RowAnnotation<T extends Record<string, unknown>> {
+  id: string;
+  rows: number[];
+  label: string;
+  className?: string;
+  metadata?: T;
+}
+export interface ColAnnotation<T extends Record<string, unknown>> {
+  id: string;
+  cols: number[];
+  label: string;
+  className?: string;
+  metadata?: T;
+}
+
+export interface PlateProps<
+  RowMetaT extends Record<string, unknown>,
+  ColMetaT extends Record<string, unknown>,
+  WellMetaT extends Record<string, unknown>,
+> {
   wells: 24 | 96 | 48 | 384;
-  rowAnnotations?: RowAnnotation[];
+  rowAnnotations?: RowAnnotation<RowMetaT>[];
+  colAnnotations?: ColAnnotation<ColMetaT>[];
+  wellAnnotations?: WellAnnotation<WellMetaT>[];
   selection: PlateSelection | null;
   setSelection: (selection: PlateSelection | null) => void;
   className?: string;
   selectionTolerance?: number;
 }
 
-export const Plate = ({
+export const Plate = <
+  RowMetaT extends Record<string, unknown>,
+  ColMetaT extends Record<string, unknown>,
+  WellMetaT extends Record<string, unknown>,
+>({
   wells,
   className,
   rowAnnotations,
+  colAnnotations,
+  wellAnnotations,
   selection,
   setSelection,
   selectionTolerance = 20,
-}: PlateProps) => {
+}: PlateProps<RowMetaT, ColMetaT, WellMetaT>) => {
```

Comment:
I'm a bit confused here about what you are buying with these generics. Is the idea that the user can provide their own row types that implement these base classes? I think I will need to do some more reading because this is next level typescript typing, but do you have a use case in mind here which is going to make the calling applications life a lot better?

Reply:
This is really good pushback - let's chat about this on our next call!